### ISSUE TITLE
Feature Request: Typescript definitions

### ISSUE THREAD

**Issue by DanHarman** [DevExpress/devextreme-reactive#386]
Could you consider offering typescript support as this is a pretty key feature for many large projects.

**Comment by kvet** [comment]
Hi,
 
Thank you for your suggestion. We'll consider adding TypeScript definitions.

**Comment by QuantumInformation** [comment]
Any news for this?

![image](https://user-images.githubusercontent.com/216566/33180099-8a7e440e-d063-11e7-85a6-23a2e6d9695b.png)

**Comment by kvet** [comment]
Hi,
 
We haven't added TypeScript definitions yet. But to be honest, we have no plans to implement it soon. We are currently working on a stable release of React Grid. We will update this ticket once we have any news.

**Comment by QuantumInformation** [comment]
thx 4 the update, note if you write the lib in TS, you get the definitions for free.

**Comment by QuantumInformation** [comment]
Adding ts defs, will likely make this lib usable for 50%+ more of developers these days

**Comment by DanHarman** [comment]
The other point is that its actually quite hard to pull in libraries without typings if you use the default spa app templates that Microsoft ship with. I think its a great idea, especially as the info seems to already be in the docs - I created a bunch of the typings myself for components I was configuring from them.

Dan Harman
________________________________

**Comment by DanHarman** [comment]
This is really patchy in coverage, but if anyone else wants to add to it then this is a decent starting point for the typings. The Filtering State is the only one I've done properly at the moment by declaring it as a react component.

```
/**
 * Typings for @devexpress/dx-react-grid since not currently available from vendor.
 */

declare module '@devexpress/dx-react-grid' {
    import * as React from 'react';

    export const DataTypeProvider: any;
    export const EditingState: any;
    export const GroupingState: any;
    export const LocalFiltering: any;
    export const LocalGrouping: any;
    export const LocalSorting: any;
    export const SelectionState: any;
    export const SortingState: any;
    export const TableColumnResizing: any;
    export const FilterCellArgs: any;

    export interface Column {
        name: string;
        getCellValue: string;
    }

    export interface Filter {
        columnName: string;
        value?: string;
    }

    export interface FilteringStateProps {
        filters?: Filter[];
        defaultFilters: Filter[];
        onFiltersChange?: (filters: Filter[]) => void;
    }

    export class FilteringState extends React.Component<FilteringStateProps> { }

    export interface FilterCellArgs {
        filter: Filter;
        setFilter: (filter: Filter) => void;
        column: Column;
        style: any;
        tableColumn: TableColumn;
    }

    export interface TableColumn {
        /** A table column’s unique identifier. */
        key: string;
        /** Specifies the table column type. The specified value affects which cell template is used to render the column. */
        type: string;
        /** Specifies the associated user column. */
        column?: Column;
        /** Specifies the table column width. */
        width?: number;
    }

    export interface TableRow {
        /** A table row’s unique identifier. */
        key: string;
        /** Specifies the table row type. The specified value affects which cell template is used to render the row. */
        type: string;
        /** Specifies the associated row’s ID. */
        rowId?: number | string;
        /** Specifies the associated row. */
        row?: any;
        /** Specifies the table row height. */
        height?: number;
    }

    export interface TableRowArgs {
        /** A table row. */
        tableRow: TableRow;
        /** A React element used to render a table row. */
        children: any;
        /** Styles that should be applied to the root row element. */
        style?: any;
    }
}
```

**Comment by Jack-Works** [comment]
@DanHarman Not quite accurate... Too many any in it.
Like `style?: any` should be `style?: React.CSSProperties`

**Comment by DanHarman** [comment]
Yeah it was a quick hack for the bits I needed. Feel free to enhance and share - it is just a starting point. To be honest I was hoping devexpress would finish it, as really not a huge amount of work and they now have an example of what it would look like...

Dan Harman
________________________________

**Comment by Jack-Works** [comment]
Here's currently I'm using
```typescript
declare module '@devexpress/dx-react-grid' {
    import React from 'react'
    /** Column of grid */
    export interface Column<Data> {
        name?: keyof Data,
        title: any,
        getCellValue?: (row: Data) => any,
        dataType?: string
    }
    //#region DataTypeProvider
    export interface DataTypeProviderProps<Col, Row = any> {
        type: string,
        formatterTemplate: (data: { column: Column<Col>, row: Row, value: any }) => React.ReactNode
    }
    class DataTypeProvider<Data = any> extends React.Component<DataTypeProviderProps<Data>> { }
    //#endregion

    class FilteringState extends React.Component<{}> { }
    class RowDetailState extends React.Component<{}> { }

    //#region PagingState
    export interface PagingStateProps {
        currentPage: number
        onCurrentPageChange: (newPage: number) => void
        pageSize: number
        totalCount: number
    }
    class PagingState extends React.Component<PagingStateProps> { }
    //#endregion

    export {
        DataTypeProvider,
        FilteringState,
        RowDetailState,
        PagingState
    }
}
```

What about find a time to merge with @DanHarman then publish it on DefinitelyTyped?

**Comment by DanHarman** [comment]
I think it would need to be a lot better  than my effort to go on definitely typed! Also since this is a commercial component, which we are happy to pay for, I’m somewhat reluctant to take on a support borden around it!

It can’t be more than 0.5-1d work to sort this, so come on devexpress!

Dan Harman
________________________________

**Comment by QuantumInformation** [comment]
Has anyone had success with these definitions yet? I'm tempted to look at this project again.

**Comment by franklixuefei** [comment]
I'm also waiting for the type definitions to roll out. This is very important as this feature may onboard so many other typescript developers. Please consider adding the index.d.ts soon :-)

**Comment by franklixuefei** [comment]
@kvet since the typescript feature has been merged, is this issue resolved?

**Comment by kvet** [comment]
Hi @franklixuefei,

We will close the issue with the feature release.

**Comment by franklixuefei** [comment]
Thanks! Any ETA on when the next release will be?

**Comment by kvet** [comment]
We plan to publish the first pre-release version next week.

**Comment by franklixuefei** [comment]
It seems like with 1.0.1 release, the typescript feature is not included. Is this feature supposed to be released in the next version?

**Comment by kvet** [comment]
I am afraid that the 1.1 branch was not ready for the pre-release last week. We will publish it soon.

**Comment by kvet** [comment]
We released initial TypeScript support in 1.1.0-beta.1. Your feedback is greatly appreciated.
 
You can install it with the `next` tag:
 
```npm i --save @devexpress/dx-react-core@next @devexpress/dx-react-grid@next```
```npm i --save @devexpress/dx-react-grid-bootstrap3@next```
```npm i --save @devexpress/dx-react-grid-material-ui@next```

**Comment by franklixuefei** [comment]
@kvet Thanks a lot for you and the team's effort in making this happen! I've upgraded the packages, but noticed a small glitch - 
<pre>
PS C:\az\EngSys\Performance\Cirrus\src\Portal\src\CirrusPortal> npm i --save @devexpress/dx-react-core@next @devexpress/dx-react-grid@next @devexpress/dx-react-grid-material-ui@next
npm WARN @devexpress/dx-react-grid@1.1.0-beta.1 requires a peer of @devexpress/dx-core@1.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN @devexpress/dx-react-grid@1.1.0-beta.1 requires a peer of @devexpress/dx-react-core@1.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN @devexpress/dx-react-grid-material-ui@1.1.0-beta.1 requires a peer of @devexpress/dx-grid-core@1.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN @devexpress/dx-react-grid-material-ui@1.1.0-beta.1 requires a peer of @devexpress/dx-react-core@1.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN @devexpress/dx-react-grid-material-ui@1.1.0-beta.1 requires a peer of @devexpress/dx-react-grid@1.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN @devexpress/dx-react-grid-material-ui@1.1.0-beta.1 requires a peer of material-ui@1.0.0-beta.33 but none is installed. You must install peer dependencies yourself.
npm WARN @devexpress/dx-grid-core@1.1.0-beta.1 requires a peer of @devexpress/dx-core@1.0.0 but none is installed. You must
install peer dependencies yourself.
</pre>
I wonder if the dependencies should also be updated as well?

**Comment by franklixuefei** [comment]
bump

**Comment by kvet** [comment]
@franklixuefei, thank you for reporting the issue. We've recently released the 1.1.0-beta.2 version, which contains the fix.

**Comment by SergeyAlexeev** [comment]
TypeScript definitions for React Grid are available starting with [v1.1.0](https://github.com/DevExpress/devextreme-reactive/releases/tag/v1.1.0) that is officially released.
